### PR TITLE
Tenant provisioning updates

### PR DESCRIPTION
**Pull request by elv-serban**
### Summary of Changes:

1. **Command:**
     ```bash
     export PRIVATE_KEY=XXX
    ./elv-live tenant_provision <tenant_contract_id> --help

    tenant_provision <tenant> [options]
    
    Provisions a new tenant account with standard media libraries and content types. Note this account must be created using
    space_tenant_create.
    
    Positionals:
      tenant  Tenant ID                                                                                  [string] [required]
    
    Options:
          --version      Show version number                                                                       [boolean]
      -v, --verbose      Verbose mode                                                                              [boolean]
          --as_url       Alternate authority service URL (include '/as/' route if necessary)                        [string]
          --help         Show help                                                                                 [boolean]
          --status       Path to the JSON File for existing tenant provision config                                 [string]
          --init_config  Displays the initial provisioning config, which can be used to input existing objects     [boolean]

     ```

2. **To run the script:**
    - To run tenant provision script:
    ```bash
    ./elv-live tenant_provision iten2eQcZKr7y47hJkoS7crja61bf8E9
    ```

   - To run tenant provision script with existing objects set in a JSON file:
    ```bash
    ./elv-live tenant_provision iten2eQcZKr7y47hJkoS7crja61bf8E9 --status <json_file>
    ```
     
4. **Config file:**

   - To get initial config either from status file (--status <file>) or empty object :
     ```bash
     ./elv-live tenant_provision iten2eQcZKr7y47hJkoS7crja61bf8E9 --init_config

      Tenant Provision
      Tenant ID: iten2eQcZKr7y47hJkoS7crja61bf8E9
      verbose: undefined
      status: undefined
      init-config: true
      kmsId: ikms2CtpuikdJCHgsdA3aYqMvtwt5K8j
      {
        "base": {
          "tenantOpsKey": "",
          "contentOpsKey": "",
          "groups": {
            "tenantAdminGroupAddress": "",
            "contentAdminGroupAddress": ""
          },
          "libraries": {
            "mastersLibraryId": null,
            "mezzanineLibraryId": null,
            "propertiesLibraryId": null
          },
          "tenantTypes": {
            "titleTypeId": null,
            "titleCollectionTypeId": null,
            "masterTypeId": null,
            "permissionsTypeId": null,
            "channelTypeId": null,
            "streamTypeId": null
          }
        },
        "liveStreaming": {
          "siteId": null
        },
        "mediaWallet": {
          "liveTypes": {
            "NFT Collection": null,
            "NFT Template": null,
            "Media Wallet Drop Event Site": null,
            "Media Wallet Marketplace": null,
            "Media Wallet Tenant": null
          },
          "objects": {
            "marketplaceId": "",
            "marketplaceHash": null,
            "dropEventId": null,
            "dropEventHash": null,
            "tenantObjectId": null
          }
        }
      }
      ```
    
   - If a provision object is set to `null` or an empty string, the script will create a new one. If it's set to `"none"`, no new object will be created.

5. **Provision Methods:**
   - Each provision method (base, mediaWallet, liveStreaming) ensures that all required parameters are set. If any dependent parameter is set to "none" or is missing, the script will throw an error.
   
6. **Final Output:**
   - As the objects are created, the results are saved to the `./tenant_status.json` file by default. 
   - If `--status <file>` flag is provided, the initial config is saved as backup file in pwd and the new results are saved to  the file provided.

### CODE REVIEW

File: src/Provision.js [hunk 1]
```diff
@@ -179,99 +208,95 @@ const InitializeTenant = async ({client, kmsId, tenantId, debug=false}) => {
...
+            ]
+          }
         }
       }
-    }
-  });
+    });
+
+    await SetObjectPermissions(client, t.tenantTypes.streamTypeId, t.groups.tenantAdminGroupAddress, t.groups.contentAdminGroupAddress);
+  }
```
Comment (elv-preethi) [inline src/Provision.js:289]: we can make common method to create and set object permission:
```
// helper function to create and set permissions for content types
  const createAndSetPermissions = async (name, metadata) => {
    const typeId = await client.CreateContentType({ name, metadata });
    await SetObjectPermissions(client, typeId, t.groups.tenantAdminGroupAddress, t.groups.contentAdminGroupAddress);
    return typeId;
  };

  // Create and set permissions for tenant types
  if (!t.tenantTypes.titleTypeId) {
    t.tenantTypes.titleTypeId = await createAndSetPermissions(`${t.tenantName} - Title`, typeMetadata);
  }

  if (!t.tenantTypes.titleCollectionTypeId) {
    t.tenantTypes.titleCollectionTypeId = await createAndSetPermissions(`${t.tenantName} - Title Collection`, typeMetadata);
  }

  if (!t.tenantTypes.masterTypeId) {
    t.tenantTypes.masterTypeId = await createAndSetPermissions(`${t.tenantName} - Title Master`, masterMetadata);
  }

  if (!t.tenantTypes.permissionsTypeId) {
    t.tenantTypes.permissionsTypeId = await createAndSetPermissions(`${t.tenantName} - Permissions`, { ...typeMetadata, public: { "eluv.manageApp": "avails-manager" } });
  }

  if (!t.tenantTypes.channelTypeId) {
    t.tenantTypes.channelTypeId = await createAndSetPermissions(`${t.tenantName} - Channel`, {
      ...typeMetadata,
      public: {
        ...typeMetadata.public,
        title_configuration: {
          "controls": ["credits", "playlists", "gallery", "channel"],
        },
      },
    });
  }

  if (!t.tenantTypes.streamTypeId) {
    t.tenantTypes.streamTypeId = await createAndSetPermissions(`${t.tenantName} - Live Stream`, {
      bitcode_flags: "playout_live",
      bitcode_format: "builtin",
      public: {
        ...typeMetadata.public,
        title_configuration: {
          "controls": ["credits", "playlists", "gallery", "live_stream"],
        },
      },
    });
  }
```

File: src/Provision.js [hunk 2]
```diff
@@ -281,211 +306,194 @@ const InitializeTenant = async ({client, kmsId, tenantId, debug=false}) => {
       }
     });
 
-    await SetObjectPermissions(client, typeId, tenantAdminGroupAddress, contentAdminGroupAddress, contentViewersGroupAddress);
+    await SetObjectPermissions(client, typeId, t.groups.tenantAdminGroupAddress, t.groups.contentAdminGroupAddress);
```
Comment (elv-preethi) [inline src/Provision.js:309]:
```
const typeId = await createAndSetPermissions(`${t.tenantName} - ${liveTypes[i].name}`, {
      ...typeMetadata,
      public: {
        ...typeMetadata.public,
        title_configuration: liveTypes[i].spec
      }
    });
```

File: src/Provision.js [hunk 3]
```diff
@@ -281,211 +306,194 @@ const InitializeTenant = async ({client, kmsId, tenantId, debug=false}) => {
...
 
-  // NFT Templates library not used for now
-  /*
-  const nftLibraryId = await client.CreateContentLibrary({
-    name: `${tenantName} - NFT Templates`,
-    kmsId,
-    metadata: STANDARD_DRM_CERT
-  });
+    await SetLibraryPermissions(client, t.libraries.mezzanineLibraryId, t.groups.tenantAdminGroupAddress, t.groups.contentAdminGroupAddress);
+  }
```
Comment (elv-preethi) [inline src/Provision.js:347]:
```
// Create libraries and set permissions
  const createLibrary = async (name, metadata) => {
    const libraryId = await client.CreateContentLibrary({ name, kmsId, metadata });
    await SetLibraryPermissions(client, libraryId, t.groups.tenantAdminGroupAddress, t.groups.contentAdminGroupAddress);
    return libraryId;
  };

  if (!t.libraries.propertiesLibraryId) {
    t.libraries.propertiesLibraryId = await createLibrary(`${t.tenantName} - Properties`);
  }

  if (!t.libraries.mastersLibraryId) {
    t.libraries.mastersLibraryId = await createLibrary(`${t.tenantName} - Title Masters`);
  }

  if (!t.libraries.mezzanineLibraryId) {
    t.libraries.mezzanineLibraryId = await createLibrary(`${t.tenantName} - Title Mezzanines`, STANDARD_DRM_CERT);
  }
```

**Review by elv-preethi — COMMENTED** [review]
looks good! 
We can avoid repetitive code.

File: src/Provision.js [hunk 4]
```diff
@@ -29,41 +34,108 @@ const STANDARD_DRM_CERT={
...
+  fs.writeFileSync(OUTPUT_FILE, JSON.stringify(config, null, 2));
+};
+
+const checkSignerTenantAccess = async ({client, tenantId}) => {
+  let tenantAdminId = await client.userProfileClient.TenantId();
+  let tenantContractId = await client.userProfileClient.TenantContractId();
+
+  if (!tenantAdminId) {
+    throw Error("No tenant admin group set for account.");
+  }
```
Comment (elv-serban) [inline src/Provision.js:72]: This is not required in the future - it's just vestigial  Let's not check for it

We require:
- user has tenantContractId and it is the same as the tenantId supplied in the command
- tenant has to have group 'tenant admins'
- user is a member of tenant admins of that tenant

File: src/Provision.js [hunk 5]
```diff
@@ -3,10 +3,15 @@ const fs = require("fs");
...
 const { ElvUtils } = require("../src/Utils");
 const Utils = require("@eluvio/elv-client-js/src/Utils.js");
+const { ElvAccount } = require("../src/ElvAccount.js");
 
-const TYPE_LIVE_DROP_EVENT_SITE = "Eluvio LIVE Drop Event Site";
-const TYPE_LIVE_TENANT = "Eluvio LIVE Tenant";
-const TYPE_LIVE_MARKETPLACE = "Eluvio LIVE Marketplace";
+const { TenantProvisionConfig } = require("./TenantProvisionConfig.js");
+const {Config} = require("./Config");
+const t = TenantProvisionConfig.tenantProvisionConfig;
```
Comment (elv-serban) [inline src/Provision.js:10]: Let's not make this global - that way the functions could also be used in other tools.  We can add it as an argument to all functions that need it.  Make it the first argument

**Comment by elv-preethi** [comment]
@elv-serban Updated the command as per the comments: https://github.com/eluv-io/elv-live-js/pull/185#issue-2462015268

File: utilities/EluvioLiveCli.js [hunk 6]
```diff
@@ -2962,13 +2973,21 @@ yargs(hideBin(process.argv))
   )
 
   .command(
-    "tenant_provision <tenant>",
+    "tenant_provision <tenant> [options]",
```
Comment (elv-serban) [inline utilities/EluvioLiveCli.js:2976]: Let's change the description
```
Provisions a new tenancy with standard media libraries and content types.  Run as the tenant root-key, as created by space_tenant_create.  This is a multi-step operation and intermediate status is saved in the local directory in the file tenant_status.json.  The operation can be resumed by specifying a status file (which indicates what operations have been executed already).
```

**Review by elv-serban — COMMENTED** [review]
Looks good!

**Review by elv-serban — COMMENTED** [review]
Looks good!

**Pull request merged**